File: SimpleEngine/fse/Pipeline.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SimpleEngine.fse
{
    public class Pipeline
    {
        private List<Provider> providers = new List<Provider>();

        public static Pipeline Create(string path, MediaLoader ml, int width, int height)
        {
            Linker linker = new Linker();
            string t = linker.read(path, ml, width, height);
            linker.link();
            Pipeline pp = linker.getPipeline( linker.getTarget(t) );
            Binder bind = new Binder(ml.FS);
            pp.bind(bind);
            bind.createBuffers();
            return pp;
        }

        private void bind(Binder binder)
        {
            foreach (Provider p in providers)
            {
                p.bind(binder);
                if (p.Target != null)
                {
                    p.Target.bind(binder);
                }
            }
        }

        internal Pipeline()
        {
        }

        public void render(RenderArgs ra)
        {
            foreach (Provider p in providers)
            {
                p.provide(ra);
            }
        }

        internal void add(Provider pr)
        {
            foreach (Provider p in pr.Providers)
            {
                add(p);
            }
            
            if (false == providers.Contains(pr))
            {
                providers.Add(pr);
            }
        }
    }
}

File: src/engine/fse/pipeline.cc
#include "engine/fse/pipeline.h"

#include <algorithm>

#include "engine/fse/binder.h"
#include "engine/fse/linker.h"
#include "engine/fse/target.h"
#include "engine/medialoader.h"

namespace simple::fse
{
    std::shared_ptr<Pipeline> Pipeline::LoadFromFile(const std::string& path, MediaLoader* ml, int width, int height)
    {
        Linker linker;
        std::string t = linker.ReadPipelineFromFile(path, ml, width, height);
        linker.Link();
        auto pp = linker.GetPipeline(linker.GetTarget(t));
        auto bind = Binder(ml->FS());
        pp->Bind(&bind);
        bind.CreateBuffers();
        return pp;
    }

    void Pipeline::Bind(Binder* binder)
    {
        for (auto p : providers)
        {
            p->Bind(binder);
            if (p->target != nullptr)
            {
                p->target->Bind(binder);
            }
        }
    }

    void Pipeline::Render(RenderArgs* ra)
    {
        OnSize(ra->width, ra->height);
        for (auto p : providers)
        {
            p->Provide(ra);
        }
    }

    void Pipeline::Add(std::shared_ptr<Provider> pr)
    {
        for (auto p : pr->providers)
        {
            Add(p);
        }

        if (std::find(providers.begin(), providers.end(), pr) == providers.end())
        {
            providers.emplace_back(pr);
        }
    }

    void Pipeline::OnSize(int width, int height)
    {
        for (auto p : providers)
        {
            p->OnSize(width, height);
        }
    }
}
